argon2/src/lib.rs

```diff
@@ -0,0 +1,901 @@
+                blake2_inner!($v1, $v5, $v9, $v13);
+                blake2_inner!($v2, $v6, $v10, $v14);
+                blake2_inner!($v3, $v7, $v11, $v15);
+                blake2_inner!($v0, $v5, $v10, $v15);
+                blake2_inner!($v1, $v6, $v11, $v12);
+                blake2_inner!($v2, $v7, $v8, $v13);
+                blake2_inner!($v3, $v4, $v9, $v14);
+            };
+        }
+
+        macro_rules! blake2_inner {
+            ($a:expr, $b:expr, $c:expr, $d:expr) => {
+                $a = fBlaMka($a, $b);
+                $d = rotr64($d ^ $a, 32);
+                $c = fBlaMka($c, $d);
+                $b = rotr64($b ^ $c, 24);
+                $a = fBlaMka($a, $b);
+                $d = rotr64($d ^ $a, 16);
+                $c = fBlaMka($c, $d);
+                $b = rotr64($b ^ $c, 63);
+            };
+        }
+
+        // Apply Blake2 on columns of 64-bit words: (0, 1, ..., 15), then
+        // (16, 17, ..31)... finally (112, 113, ...127)
+        for i in 0..8 {
+            blake2_round!(
+                blockR.0[16 * i],
+                blockR.0[16 * i + 1],
+                blockR.0[16 * i + 2],
+                blockR.0[16 * i + 3],
+                blockR.0[16 * i + 4],
+                blockR.0[16 * i + 5],
+                blockR.0[16 * i + 6],
+                blockR.0[16 * i + 7],
+                blockR.0[16 * i + 8],
+                blockR.0[16 * i + 9],
+                blockR.0[16 * i + 10],
+                blockR.0[16 * i + 11],
+                blockR.0[16 * i + 12],
+                blockR.0[16 * i + 13],
+                blockR.0[16 * i + 14],
+                blockR.0[16 * i + 15]
+            );
+        }
+
+        // Apply Blake2 on rows of 64-bit words: (0, 1, 16, 17, ...112, 113), then
+        // (2, 3, 18, 19, ..., 114, 115).. finally (14, 15, 30, 31, ..., 126, 127)
+        for i in 0..8 {
+            blake2_round!(
+                blockR.0[2 * i],
+                blockR.0[2 * i + 1],
+                blockR.0[2 * i + 16],
+                blockR.0[2 * i + 17],
+                blockR.0[2 * i + 32],
+                blockR.0[2 * i + 33],
+                blockR.0[2 * i + 48],
+                blockR.0[2 * i + 49],
+                blockR.0[2 * i + 64],
+                blockR.0[2 * i + 65],
+                blockR.0[2 * i + 80],
+                blockR.0[2 * i + 81],
+                blockR.0[2 * i + 96],
+                blockR.0[2 * i + 97],
+                blockR.0[2 * i + 112],
+                blockR.0[2 * i + 113]
+            );
+        }
+
+        // TODO(tarcieri): get rid of this with `Copy`
+        self.copy_block(&block_tmp);
+        self.xor_block(&blockR);
+    }
+}
+
+/// designed by the Lyra PHC team
+fn fBlaMka(x: u64, y: u64) -> u64 {
+    let m = 0xFFFFFFFF;
+    let xy = Wrapping((x & m) * (y & m)) * Wrapping(2);
+    (Wrapping(x) + Wrapping(y) + xy).0
+}
+
+fn rotr64(w: u64, c: u64) -> u64 {
+    (w >> c) | (w << (64 - c))
+}
```

Comment:
@newpavlov any ideas on how we could leverage the `blake2` crate for this?

Reply:
Not sure, I will have to refresh my memory about BLAKE2 and familiarize myself with the Argon2 algorithm first.

Also I think that `blake2` rework is a long overdue. We probably should do it based on the `blake2b_simd` and `blake2s_simd` crates and use this chance to expose parts necessary for Argon2.

Reply:
Cool, well I guess we'll need to use something like this (I'll clean it up a bit more) but it'd be great to be able to leverage a SIMD implementation here